README

POPS*: Parameter OPtimised Surface of proteins and nucleic acids 
----------------------------------------------------------------
Calculation of solvent accessible surface areas (SASAs) of biopolymers,
currently proteins and nucleic acids.


References
----------
Users publishing results obtained with the program and its applications
should acknowledge its use by the following citation:

- Implicit solvent
  Fraternali, F. and van Gunsteren, W.F.
  An efficient mean solvation force model for use in molecular 
  dynamics simulations of proteins in aqueous solution.
  Journal of Molecular Biology 256 (1996) 939-948.

- POPS* method
  Fraternali, F. and Cavallo, L.
  Parameter optimized surfaces (POPS*): analysis of key interactions
  and conformational changes in the ribosome.
  Nucleic Acids Research 30 (2002) 2950-2960.

- POPS* server
  Cavallo, L., Kleinjung, J. and Fraternali, F.
  POPS: A fast algorithm for solvent accessible surface areas 
  at atomic and residue level.
  Nucleic Acids Research 31 (2003) 3364-3366.

- POPSCOMP server
  Kleinjung, J. and Fraternali, F.
  POPSCOMP: an automated interaction analysis of biomolecular complexes.
  Nucleic Acids Research 33 (2005) W342-W346.


Install / Uninstall
-------------------
Run the 'bootstrap' shell script and follow the
general 'INSTALL' instructions.

In case the Autoconf version is too old, run
% autoconf --version
and use the result to replace the version number in configure.ac
AC_PREREQ([2.59])
to reconfigure with the bootstrap script.

The configuration option '--enable-debug' creates a debuggable binary,
the option '--enable-profiling' enables profiling with 'gprof'.

For automatic code documentation execute 'doxygen doxygen.cfg'
in the 'src' directory.
Documentation files are created in 'doc/html' and 'doc/latex'.
The latex documentation is completed by executing 'make pdf' in the
'doc/latex' directory, which creates 'refman.ps' and 'refman.pdf'.


Usage
-----
pops [--pdb ...] [OPTIONS ...]
	 INPUT OPTIONS
	   --pdb <PDB input>		(mode: mandatory, type: char  , default: void)
	   --traj <trajectory input>	(mode: optional, type: char  , default: void)
	 MODE OPTIONS
	   --coarse			(mode: optional , type: no_arg, default: off)
	   --multiModel (disabled)	(mode: optional , type: no_arg, default: off)
	   --rProbe <probe radius [A]>	(mode: optional , type: float , default: 1.4)
	   --silent			(mode: optional , type: no_arg, default: off)
	 OUTPUT OPTIONS
	   --popsOut <POPS output>	(mode: optional , type: char  , default: pops.out)
	   --popstrajOut <POPS output>	(mode: optional , type: char  , default: popstraj.out)
	   --sigmaOut <SFE output>	(mode: optional , type: char  , default: sigma.out)
	   --sigmatrajOut <SFE output>	(mode: optional , type: char  , default: sigmatraj.out)
	   --compositionOut		(mode: optional , type: no_arg, default: off)
	   --typeOut			(mode: optional , type: no_arg, default: off)
	   --topologyOut		(mode: optional , type: no_arg, default: off)
	   --atomOut			(mode: optional , type: no_arg, default: off)
	   --residueOut			(mode: optional , type: no_arg, default: off)
	   --chainOut			(mode: optional , type: no_arg, default: off)
	   --neighbourOut		(mode: optional , type: no_arg, default: off)
	   --parameterOut		(mode: optional , type: no_arg, default: off)
	   --noTotalOut			(mode: optional , type: no_arg, default: off)
	   --noHeaderOut		(mode: optional , type: no_arg, default: off)
	   --padding			(mode: optional , type: no_arg, default: off)
	 INFO OPTIONS
	   --cite			(mode: optional , type: no_arg, default: off)
	   --version			(mode: optional , type: no_arg, default: off)
	   --help


Program Design
--------------
- PDB input file (read input) 
- Groups (assign atom group id)
- Types (assign GROMOS atom type id)
- Topology (compute molecular topology)
- SASA (compute SASA values at atoms, chain and molecule level)
- Output (print output)

= Determine the atom and residue types according to 'sasa_data.h' ('type'),
Possible complications in the handling of PDB files and the way it
is dealt with by POPSc are listed here.
- Hydrogen atoms : Hydrogen atoms are skipped.
- Alternative locations : Alternative locations are skipped.
- Multiple chains : The chain number of each atom is printed in the 
	'atomOut' output. Using the option 'chainOut' prints the
	chain areas.
- Multiple models : Only the first model is used.
- Missing atoms/residues : No detection of missing atoms/residues.
- Unrecognised standard atoms/residues (ATOM entries): 
	The program issues a warning and/or exits.
	Some programs create non-standard atom entries, which are correct within
	the limits of the PDB format definition, but differ from the standard
	format that is used by the PDB for amino acids and nucleotides.
	POPS expects the standard format as in these examples:
* amino acid
-ATOM     41  CA  PRO A  69      -3.665   2.182  -6.278  1.00  0.00           C
* nucleotide
-ATOM     33  P     G A  17       4.130  21.140  -1.230  1.00  0.00           P 
- Hetero atoms (HETATM entries): These are treated differently from the
standard ATOM entries. Atom names are matched against the standard
atom names [' N  ','  CA',' C  ',' O  '] to recognise modified
amino acids, otherwise atom names are matched against the elements
['C','N','O','P','S'], in which case these atoms are renamed to
['C_','N_','O_','P_','S_'] to indicate their generic nature.
Unrecognised HETATM entries are skipped!
- For a complete list of residues see:
ftp://ftp.wwpdb.org/pub/pdb/data/monomers/

= Atomgroup assignment (code file 'atomgroup').
POPS* uses a specific sigma parameter for each atom type.
Atom types with the same sigma can be grouped together, 
yielding a group-specific sigma parameter.

= Type assignment for atoms and residues (code file 'type').
The data structure that defines the atom and residue types is specified
as data type 'ConstantSasa' in 'sasa_const.h', 
while the actual parametrisation is kept in 'sasa_data.h'.
'sasa_data.h' defines an array 'ConstantSasa constant_sasa_data[]'.
Each element of this array is a complete POPS parametrisation.
At the moment there are two parametrisations, element '0' is the
atomic level, element '1' is the coarse-grained residue level.
The pointer 'constant_sasa' in the main program ('pops') points
to either element, which defines the parametrisation to be used.
Other parametrisations can be defined by appending new elements
to the 'constant_sasa_data[]' array.

= Determine the molecular Topol (code file 'topol').
The determination of the molecular topology proceeds progressively
from 1-2 interactions (bonds) over 1-3 interactions (angles) to
1-4 interactions (torsions). Any other pairwise atom pair relationship 
is by definition non-bonded. The 1-(i+1) interactions are assembled
from all pariwise combinations of 1-i interaction by testing identical
atom overlaps. The determination of ring topologies is a 
painfully slow and error-prone procedure if not implemented ingeniously.
Therefore, the ring status of each atom (0=no ring, 1=ring) has been
included in the 'ConstantSasa' data type so that there is no need to
determine it. This means the treatment of cyclic HETATM molecules
requires specific attention.

= SASA: Compute AtomSasa, ResSasa, ChainSasa and MolSasa (code file 'sasa').
The SASA calculation has been broken down into elementary equations (as
inline code) and a progressive strategy to approximate atom SASAs.
Starting from the SASA of the isolated atom, overlap areas are subtracted
progressively for all 1-(i+n) interactions. Atomic SASAs are summed up
to yield residue, chain and molecule SASAs.

= Output
Flags controlling the amount and type of output.
The default ouput is the overall SASA of the molecule, including the
hydrophilic and hydrophobic contributions. More output can be produced
with the '--*Out' command line flags.

ATOM SASAs
AtomNr : atom number in molecular coordinate file
AtomNe : atom name in molecular coordinate file
ResidNe : residue name in molecular coordinate file
Chain : chain name in molecular coordinate file (can be void)
ResidNr : residue number in molecular coordinate file
SASA/A^2 : solvent accessible surace area in Angstrom^2 units
Q(SASA) : quotient of SASA and Surf (below), i.e. the fraction of SASA
N(overl) : number of overlaps with atom neighbours
AtomTp : atom type code (GROMOS van der Waals atom type)
AtomGp : atom group code: positive=1, negative=2, polar=3, aromatic=4, aliphatic=5
Surf/A^2 : surface area of isolated atom

RESIDUE SASAs
ResidNe : residue name in molecular coordinate file
Chain : chain name in molecular coordinate file (can be void)
ResidNr : residue number in molecular coordinate file
Phob/A^2 : hydrophobic solvent accessible surace area in Angstrom^2 units
Phil/A^2 : hydrophilic solvent accessible surace area in Angstrom^2 units
Total/A^2 : total solvent accessible surace area in Angstrom^2 units
Q(SASA) : quotient of SASA and Surf (below), i.e. the fraction of SASA
N(overl) : number of overlaps with residue neighbours
Surf/A^2 : surface area of isolated residue

CHAIN SASAs
Chain : chain number 
Id : chain name in molecular coordinate file (can be void)
AtomRange : range of atom numbers in chain
ResidRange : range of residue numbers in chain
Phob/A^2 : hydrophobic solvent accessible surace area in Angstrom^2 units
Phil/A^2 : hydrophilic solvent accessible surace area in Angstrom^2 units
Total/A^2 : total solvent accessible surace area in Angstrom^2 units

MOLECULE SASAs
Phob/A^2 : hydrophobic solvent accessible surace area in Angstrom^2 units
Phil/A^2 : hydrophilic solvent accessible surace area in Angstrom^2 units
Total/A^2 : total solvent accessible surace area in Angstrom^2 units

bSASAs
The same type of information as for the SASAs is produced for bSASAs,
which are the buried SASAs of side-chain atoms only.
bSASAs serve as a component in the definition of a structural 
environment (3D profile).
Phob/A^2 : solvent accessible surface area buried by hydrophobic neighbour
			atoms in Angstrom^2 units
Phil/A^2 : solvent accessible surface area buried by hydrophilic neighbour
			atoms in Angstrom^2 units
Total/A^2 : solvent accessible surface area buried by all neighbour
			atoms in Angstrom^2 units

INTERFACE
Information of interface residues in complexes can be obtained with switch
--interfaceOut. For each atom of the input structure, the nearest neighbour
is computed based on the neighbour list and printed if the two contact
atoms are in different chains. Note that this pairlist is not symmetric:
atom distance A-B might be shortest for A, but distance B-C might be shortest
for B. Also, the list might change with protein motion. Therefore it depends
on the interpretation to be derived from the interface whether it is suitable
to use the atom pairs, the residue pairs and/or their distances.


= Background Information
The area equation is defined by a product ∏ of terms that estimate
the reduction of SASA of atom i by the overlap with its 
neighbours j (Hasel at al., 1988):
∏^N_i=1 [1 - (p_i p_ij b_ij (r_ij) / S_i)].
i is the atom for which the POPS area is computed, j is any of N neighbour atoms.
p_i  is an atom type specific SASA parameter.
p_ij is a sphere overlap parameter depending on the degree of bonding between i and j (1-2, 1-3, 1-4, 1-5, ...).
b_ij is a geometric construct based on the radii and distance (r_ij) of i and j.
S_i is the SASA of the free atom i (no neighbours).

The atom specific parameters (radii, SASAs) are listed on the web server's
<a href="#Parameters">Parameters</a> files for the atoms of all standard PDB residues,   
followed by the connectivity parameters (p<sub>ij</sub>, b<sub>ij</sub>) and the
solvent radius for water.


Return values
-------------
0 : Clean termination.
1 : Error:
- Input error (wrong file name).
- Too few arguments.
- Unknown standard residue or atom. 
- Too short atom distance.
- Less than 2 atoms, 2 bonds, 2 angles or 2 torsions in input structure.
- Assertions: Internal data flow is partially controlled by assertions.
  The 'assert' macro does not return an error value itself,
  but the program exits with a (non-zero) error code and a comment.
- Warning messages: Error and warninig output is directed to 'stderr'.
- Progress messsages: Progress output is sparse and directed to 'stdout'.
- Default result output: By default the program prints results to 'pops.out'.


Servers
-------
POPS: http://mathbio.crick.ac.uk/wiki/POPS
POPSCOMP: http://mathbio.crick.ac.uk/wiki/POPSCOMP


Code
----
https://github.com/jkleinj/POPS


Releases
--------
https://github.com/jkleinj/POPS/releases
see ChangeLog for the release history


Contact
-------
franca.fraternali@kcl.ac.uk
jens@jkleinj.eu


-------------------------------------------------------------------------------
Copyright (C) 2002-2017 Franca Fraternali (program author)
Copyright (C) 2008-2017 Jens Kleinjung (modular C code)
Copyright (C) 2002 Kuang Lin and Valerie Hindie (translation to C)
Copyright (C) 2002 Luigi Cavallo (parametrisation)

Availability
------------
The program is made available under the GNU Public License for academic
scientific purposes, under the condition that proper acknowledgement
is made to the authors of the program in publications resulting from the use
of the program.

License
-------
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.
This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.
You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.



NEWS
- R-compatible output
- inclusion of the POPSCOMP method into the POPS*
- gzip library
- XML parser (libxml2 library)
- JSON output
- computation of solvation free energies
- git revision control
- Compilation/check/installation on any POSIX platform with GNU Build Tools.
- POPSc is accelerated by hard coding of ring topologies.



src/arg.c
/*==============================================================================
arg.c : parse command line arguments
Copyright (C) 2007 Jens Kleinjung
Read the COPYING file for license information.
==============================================================================*/

#include "config.h"
#include "arg.h"

#ifdef MPI
#include <mpi.h>
#endif
extern int nodes;
extern int my_rank;

/*____________________________________________________________________________*/
/** print version */
static void print_version()
{
	
	fprintf(stdout, "\n"
					" __   __   __   __   *\n"
					"|__) /  \\ |__) /__` \n"
					"|    \\__/ |    .__/\n"); 

	fprintf(stdout, "\n%s %s\n", PACKAGE, VERSION);
}

/*____________________________________________________________________________*/
/** print header */
static void print_header()
{
    fprintf(stdout, "\nPOPS* : Parameter OPtimised Surface of proteins and nucleic acids\n");
}

/*____________________________________________________________________________*/
/** print license */
static void print_license()
{
    fprintf(stdout, "\nCopyright (C) 2002-2018 Franca Fraternali (program author)\n"
			"Copyright (C) 2008-2018 Jens Kleinjung (modular C code)\n"
			"Copyright (C) 2002 Kuang Lin and Valerie Hindie (translation to C)\n"
			"Copyright (C) 2002 Luigi Cavallo (parametrisation)\n"
			"POPS* is free software and comes with ABSOLUTELY NO WARRANTY.\n"
			"You are welcome to redistribute it under certain conditions.\n"
			"Read the COPYING file for distribution details.\n\n");
}

/*____________________________________________________________________________*/
/** print citation */
static void print_citation()
{
	fprintf(stdout, "\nPOPS* METHOD\n");
    fprintf(stdout, "Fraternali, F. and van Gunsteren, W.F.\n"
			"An efficient mean solvation force model for use in molecular"
			" dynamics simulations of proteins in aqueous solution.\n"
			"Journal of Molecular Biology 256 (1996) 939-948.\n");
    fprintf(stdout, "\nFraternali, F. and Cavallo, L.\n"
			"Parameter optimized surfaces (POPS): analysis of key interactions "
			"and conformational changes in the ribosome.\n"
			"Nucleic Acids Research 30 (2002) 2950-2960.\n");
	fprintf(stdout, "\nPOPS SERVER\n");
	fprintf(stdout, "Cavallo, L., Kleinjung, J. and Fraternali, F.\n"
			"POPS: A fast algorithm for solvent accessible surface areas "
			"at atomic and residue level.\n"
			"Nucleic Acids Research 31 (2003) 3364-3366.\n");
	fprintf(stdout, "\nPOPSCOMP SERVER\n");
	fprintf(stdout, "Kleinjung, J. and Fraternali, F.\n"
			"POPSCOMP: an automated interaction analysis "
			"of biomolecular complexes.\n"
			"Nucleic Acids Research 33 (2005) W342-W346.\n\n");
}

/*____________________________________________________________________________*/
/** set defaults */
static void set_defaults(Arg *arg, Argpdb *argpdb)
{
    arg->pdbInFileName = "";
    arg->pdbmlInFileName = "";
	arg->pdbml = 0;
	arg->zipped = 0;
	arg->trajInFileName = 0; /* trajectory input file */ 
	argpdb->coarse = 0; /* Calpha- or Pphosphate-only computation [0,1] */
	argpdb->hydrogens = 0; /* read hydrogens [0,1] */
	argpdb->multiModel = 0; /* read multiple models [0,1] */
	argpdb->partOcc = 0; /* partial occupancy [0,1] */
	arg->rProbe = 1.4; /* probe radius (in Angstrom) */
	arg->silent = 0; /* suppress stdout */
    arg->sasaOutFileName = "pops.out";
    arg->sasatrajOutFileName = "popstraj";
    arg->bsasaOutFileName = "popsb.out";
    arg->bsasatrajOutFileName = "popsbtraj";
	arg->compositionOut = 0; /* output of molecule composition */
    arg->sigmaOutFileName = "sigma.out";
    arg->sigmatrajOutFileName = "sigmatraj";
    arg->interfaceOutFileName = "interface.out";
	arg->interfaceOut = 0; /* output of interface residue pairs */
	arg->typeOut = 0; /* output of atom/residue types */
	arg->topologyOut = 0; /* output of molecule topology */
	arg->atomOut = 0; /* output of atom areas */
	arg->residueOut = 0; /* output of residue areas */
	arg->chainOut = 0; /* output of chain areas */
    arg->neighbourOutFileName = "neli.out";
	arg->neighbourOut = 0; /* output of neighbour list */
    arg->parameterOutFileName = "para.out";
	arg->parameterOut = 0; /* output of atom parameters (for benchmarking) */
	arg->noTotalOut = 0; /* suppress output of total area (for benchmarking) */
	arg->noHeaderOut = 0; /* suppress output headers (for benchmarking) */
	arg->padding = 0; /* add lines to pad the missing hydrogen atom lines 
						(for benchmarking) */
	arg->rout = 0; /* output for R-version of POPSCOMP */ 
	arg->jsonOut = 0; /* JSON output */
	arg->jsonOutFileName = "pops.json";
	arg->jsonbOutFileName = "popsb.json";
}

/*____________________________________________________________________________*/
/** check input */
static void check_input(Arg *arg, Argpdb *argpdb)
{
	if ((strlen(arg->pdbInFileName) == 0) &&
		(strlen(arg->pdbmlInFileName) == 0))
		Error("Invalid PDB file name");
	assert(arg->pdbml == 0 || arg->pdbml == 1);
	assert(arg->zipped == 0 || arg->zipped == 1);
	assert(argpdb->coarse == 0 || argpdb->coarse == 1);
    assert(argpdb->hydrogens == 0 || argpdb->hydrogens == 1);
	assert(argpdb->multiModel == 0 || argpdb->multiModel == 1);
	assert(argpdb->partOcc == 0 || argpdb->partOcc == 1);
	assert(arg->rProbe > 0);
	assert(strlen(arg->sasaOutFileName) > 0);
	assert(strlen(arg->sasatrajOutFileName) > 0);
	assert(strlen(arg->bsasaOutFileName) > 0);
	assert(strlen(arg->bsasatrajOutFileName) > 0);
	assert(strlen(arg->sigmaOutFileName) > 0);
	assert(strlen(arg->sigmatrajOutFileName) > 0);
	assert(strlen(arg->interfaceOutFileName) > 0);
	assert(arg->interfaceOut == 0 || arg->interfaceOut == 1);
	assert(arg->compositionOut == 0 || arg->compositionOut == 1);
	assert(arg->typeOut == 0 || arg->typeOut == 1);
	assert(arg->topologyOut == 0 || arg->topologyOut == 1);
	assert(arg->atomOut == 0 || arg->atomOut == 1);
	assert(arg->residueOut == 0 || arg->residueOut == 1);
	assert(arg->chainOut == 0 || arg->chainOut == 1);
	assert(arg->neighbourOut == 0 || arg->neighbourOut == 1);
	assert(arg->parameterOut == 0 || arg->parameterOut == 1);
	assert(arg->noTotalOut == 0 || arg->noTotalOut == 1);
	assert(arg->silent == 0 || arg->silent == 1);
	assert(arg->noHeaderOut == 0 || arg->noHeaderOut == 1);
	assert(arg->padding == 0 || arg->padding == 1);
	assert(arg->rout == 0 || arg->rout == 1);
	assert(arg->jsonOut == 0 || arg->jsonOut == 1);
}

/*____________________________________________________________________________*/
/** print command line arguments */
static void print_args(Arg *arg, Argpdb *argpdb)
{
    time_t now;
    time(&now);

    fprintf(stdout, "\ndate: %s", ctime(&now));
    fprintf(stdout, \
                    "pdb/pdbml: %s%s\n"
					"zipped: %d\n"
                    "traj: %s\n"
                    "coarse: %d\n"
                    "multiModel: %d\n"
                    "rProbe: %f\n"
					"jsonOut: %d\n",
        arg->pdbInFileName, arg->pdbmlInFileName, arg->zipped,
		arg->trajInFileName, argpdb->coarse,
        0, arg->rProbe, arg->jsonOut);
    fflush(stdout);
}

/*____________________________________________________________________________*/
/** parse command line long_options */
int parse_args(int argc, char **argv, Arg *arg, Argpdb *argpdb)
{
	int c;
	const char usage[] = "\npops [--pdb ... | --pdbml ...] [OPTIONS ...]\n\
	 INPUT OPTIONS\n\
       Input is either a *.pdb[.gz] file (--pdb) or a *.xml[.gz] file (--pdbml).\n\
	   --pdb <PDB input>\t\t(mode: optional, type: char  , default: void)\n\
	   --pdbml <PDBML input>\t(mode: optional, type: char  , default: void)\n\
	   --traj <trajectory input>\t(mode: optional , type: char  , default: void)\n\
	   --zipped\t\t\t(mode: opional , type: bool  , default: false)\n\
	 MODE OPTIONS\n\
	   --coarse\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --hydrogens\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --multiModel\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --partOcc\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --rProbe <probe radius [A]>\t(mode: optional , type: float , default: 1.4)\n\
	   --silent\t\t\t(mode: optional , type: no_arg, default: off)\n\
	 OUTPUT OPTIONS\n\
	   --popsOut <POPS output>\t(mode: optional , type: char  , default: pops.out)\n\
	   --popstrajOut <POPS output>\t(mode: optional , type: char  , default: popstraj.out)\n\
	   --popsbOut <POPSb output>\t(mode: optional , type: char  , default: popsb.out)\n\
	   --popsbtrajOut <POPSb output>(mode: optional , type: char  , default: popsbtraj.out)\n\
	   --sigmaOut <SFE output>\t(mode: optional , type: char  , default: sigma.out)\n\
	   --sigmatrajOut <SFE output>\t(mode: optional , type: char  , default: sigmatraj.out)\n\
	   --interfaceOut\t\t(mode: optional , type: no_arg, default: off)\n\
	   --compositionOut\t\t(mode: optional , type: no_arg, default: off)\n\
	   --typeOut\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --topologyOut\t\t(mode: optional , type: no_arg, default: off)\n\
	   --atomOut\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --residueOut\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --chainOut\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --neighbourOut\t\t(mode: optional , type: no_arg, default: off)\n\
	   --parameterOut\t\t(mode: optional , type: no_arg, default: off)\n\
	   --noTotalOut\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --noHeaderOut\t\t(mode: optional , type: no_arg, default: off)\n\
	   --padding\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --rout\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --jsonOut\t\t\t(mode: optional , type: no_arg, default: off)\n\
	 INFO OPTIONS\n\
	   --cite\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --version\t\t\t(mode: optional , type: no_arg, default: off)\n\
	   --help\n";

    if (argc < 2) {
		print_header();
        fprintf(stderr, "%s", usage);
		print_license();
        exit(0);
    }

    set_defaults(arg, argpdb);

    /** long option definition */
    static struct option long_options[] = {
        {"pdb", required_argument, 0, 1},
        {"traj", required_argument, 0, 2},
        {"coarse", no_argument, 0, 3},
        {"multiModel", no_argument, 0, 4},
        {"rProbe", required_argument, 0, 5},
        {"popsOut", required_argument, 0, 6},
        {"popstrajOut", required_argument, 0, 7},
        {"popsbOut", required_argument, 0, 8},
        {"popsbtrajOut", required_argument, 0, 9},
        {"sigmaOut", required_argument, 0, 10},
        {"sigmatrajOut", required_argument, 0, 11},
        {"interfaceOut", no_argument, 0, 12},
        {"silent", no_argument, 0, 13},
        {"compositionOut", no_argument, 0, 14},
        {"typeOut", no_argument, 0, 15},
        {"topologyOut", no_argument, 0, 16},
        {"atomOut", no_argument, 0, 17},
        {"residueOut", no_argument, 0, 18},
        {"chainOut", no_argument, 0, 19},
        {"neighbourOut", no_argument, 0, 20},
        {"parameterOut", no_argument, 0, 21},
        {"noTotalOut", no_argument, 0, 22},
        {"noHeaderOut", no_argument, 0, 23},
        {"padding", no_argument, 0, 24},
        {"rout", no_argument, 0, 25},
        {"jsonOut", no_argument, 0, 26},
        {"hydrogens", no_argument, 0, 27},
        {"partOcc", no_argument, 0, 28},
        {"pdbml", required_argument, 0, 29},
        {"zipped", no_argument, 0, 30},
        {"cite", no_argument, 0, 40},
        {"version", no_argument, 0, 41},
        {"help", no_argument, 0, 42},
        {0, 0, 0, 0}
    };

    /** assign parameters to long options */
    while ((c = getopt_long(argc, argv, "1:2:3 4 5:6:7:8:9:10:11:12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29:30 40 41", long_options, NULL)) != -1) {
        switch(c) {
            case 1:
                arg->pdbInFileName = optarg;
                break;
            case 2:
                arg->trajInFileName = optarg;
                break;
            case 3:
                argpdb->coarse = 1;
                break;
            case 4:
                argpdb->multiModel = 1;
                break;
            case 5:
                arg->rProbe = atof(optarg);
                break;
            case 6:
                arg->sasaOutFileName = optarg;
                break;
            case 7:
                arg->sasatrajOutFileName = optarg;
                break;
            case 8:
                arg->bsasaOutFileName = optarg;
                break;
            case 9:
                arg->bsasatrajOutFileName = optarg;
                break;
            case 10:
                arg->sigmaOutFileName = optarg;
                break;
            case 11:
                arg->sigmatrajOutFileName = optarg;
            case 12:
                arg->interfaceOut = 1;
                break;
            case 13:
                arg->silent = 1;
                break;
            case 14:
                arg->compositionOut = 1;
                break;
            case 15:
                arg->typeOut = 1;
                break;
            case 16:
                arg->topologyOut = 1;
                break;
            case 17:
                arg->atomOut = 1;
                break;
            case 18:
                arg->residueOut = 1;
                break;
            case 19:
                arg->chainOut = 1;
                break;
            case 20:
                arg->neighbourOut = 1;
                break;
            case 21:
                arg->parameterOut = 1;
                break;
            case 22:
                arg->noTotalOut = 1;
                break;
            case 23:
                arg->noHeaderOut = 1;
                break;
            case 24:
                arg->padding = 1;
                break;
            case 25:
                arg->rout = 1;
                break;
            case 26:
                arg->jsonOut = 1;
                break;
            case 27:
                argpdb->hydrogens = 1;
                break;
            case 28:
                argpdb->partOcc = 1;
                break;
            case 29:
                arg->pdbmlInFileName = optarg;
				arg->pdbml = 1;
				break;
            case 30:
				arg->zipped = 1;
				break;
            case 40:
				print_citation();
                exit(0);
            case 41:
				print_version();
				print_license();
                exit(0);
            default:
                fprintf(stderr, "%s", usage);
				print_license();
                exit(1);
        }
    }

	check_input(arg, argpdb);
    print_header();
    print_args(arg, argpdb);

    return 0;
}

